deploy.py

```diff
@@ -149,43 +154,50 @@ def run_desc_template(args):
         ApplicationName=args.app,
         TemplateName=args.template,
     )
-    pprint.PrettyPrinter(stream=sys.stderr).pprint(template)
+    json.dump(template, sys.stdout, sort_keys=True, indent=2, cls=ExtendedJSONEncoder)
+    print()
 
 
 def run_new_env(args):
-    state = safecast_deploy.state.State(
-        args.app,
-        args.env,
-        new_version=args.version,
-        new_arn=args.arn
-    )
-    safecast_deploy.new_env.NewEnv(state, not args.no_update_templates).run()
+    eb_client = boto3.client('elasticbeanstalk')
+    result_logger = ResultLogger()
+    state = safecast_deploy.state.State(args.app, eb_client)
+    config_saver = safecast_deploy.config_saver.ConfigSaver(eb_client, result_logger)
+    safecast_deploy.new_env.NewEnv(
+        EnvType(args.env),
+        state.old_aws_state,
+        state.new_aws_state(new_version=args.version),
+        boto3.client('elasticbeanstalk'),
+        result_logger,
+        config_saver,
+        (not args.no_update_templates),
+    ).run()
 
 
 def run_same_env(args):
-    state = safecast_deploy.state.State(
-        args.app,
-        args.env,
-        new_version=args.version,
-    )
-    safecast_deploy.same_env.SameEnv(state).run()
+    state = safecast_deploy.state.State(args.app, boto3.client('elasticbeanstalk'))
+    safecast_deploy.same_env.SameEnv(
+        EnvType(args.env),
+        state.old_aws_state,
+        state.new_aws_state(new_version=args.version),
```

Comment:
I got an error here trying to take it for a spin:

```
~/safecast/api/safecast_deploy refactor                                                                                                                                                               01:57:47
python-3.8.5 ❯ ./deploy.py same_env api dev api-master-1093-1baaa1ddec89e5074a19c90d9b5f782c80e4b2d3
Traceback (most recent call last):
  File "./deploy.py", line 204, in <module>
    main()
  File "./deploy.py", line 199, in main
    parse_args()
  File "./deploy.py", line 120, in parse_args
    args.func(args)
  File "./deploy.py", line 182, in run_same_env
    state.new_aws_state(new_version=args.version),
TypeError: new_aws_state() missing 1 required positional argument: 'target_env_type'
```

Maybe this should be like this?

```suggestion
        state.new_aws_state(env_type, new_version=args.version),
```

Reply:
Though I tried that then got 
```
/Users/matschaffer/safecast/api/safecast_deploy/.direnv/python-3.8.5/bin/python /Users/matschaffer/safecast/api/safecast_deploy/deploy.py same_env api dev api-master-1093-1baaa1ddec89e5074a19c90d9b5f782c80e4b2d3
Traceback (most recent call last):
  File "/Users/matschaffer/safecast/api/safecast_deploy/deploy.py", line 205, in <module>
    main()
  File "/Users/matschaffer/safecast/api/safecast_deploy/deploy.py", line 200, in main
    parse_args()
  File "/Users/matschaffer/safecast/api/safecast_deploy/deploy.py", line 120, in parse_args
    args.func(args)
  File "/Users/matschaffer/safecast/api/safecast_deploy/deploy.py", line 183, in run_same_env
    state.new_aws_state(env_type, new_version=args.version),
  File "/Users/matschaffer/safecast/api/safecast_deploy/safecast_deploy/state.py", line 30, in new_aws_state
    new_env_num = max([(tier.num + 1) % 1000 for (tier_type, tier) in old_aws_state.envs[target_env_type]], key=operator.itemgetter(1))
NameError: name 'old_aws_state' is not defined
```

So still need to dig more I guess.

Reply:
Sorry, not a lot of test coverage on that part `state.py`. I've pushed a fix to that particular issue (needed `self.` in front of `old_aws_state`) and I'm going to see if I can add a basic unit test of that code.

Reply:
Getting this now
```
    state.new_aws_state(env_type, new_version=args.version),
  File "/Users/matschaffer/safecast/api/safecast_deploy/safecast_deploy/state.py", line 30, in new_aws_state
    new_env_num = max([(tier.num + 1) % 1000 for (tier_type, tier) in self.old_aws_state.envs[target_env_type]], key=operator.itemgetter(1))
  File "/Users/matschaffer/safecast/api/safecast_deploy/safecast_deploy/state.py", line 30, in <listcomp>
    new_env_num = max([(tier.num + 1) % 1000 for (tier_type, tier) in self.old_aws_state.envs[target_env_type]], key=operator.itemgetter(1))
ValueError: too many values to unpack (expected 2)
```

Plus it looks like you might be missing `import operator`